### ISSUE TITLE
Building type quest refuses to disappear when solved

### ISSUE THREAD

**Issue by mnalis** [streetcomplete/StreetComplete#3290]
**How to Reproduce**

While solving hundreds of quests, I've come across one `building type` which refuses to disappear when answered.
See attached video. I do not know why, other quest disappear normally when solved. 
Auto-upload is disabled, and data was manually scanned earlier that day. 

It was this building: https://www.openstreetmap.org/note/2853877
I still have not uploaded the survey data (but will soon, unless people tell me to hold it for better debug):

Video:

https://user-images.githubusercontent.com/156656/133517966-13e9ded3-2578-4e74-ae47-6c9d4080d66a.mp4

**Versions affected**
Android 10, SC v34.0

**Comment by westnordost** [comment]
What if you wait?

**Comment by AlertSubject** [comment]
Would force stopping the app + clearing its cache usually fix such issues?

It's like the "have you tried restarting it" of android

**Comment by mnalis** [comment]
> What if you wait?

Yesterday I was playing with that quest for few minutes (first when it happened, then to verify, then to take video), and it persisted for that long at least. 

Today, I took the phone, and the quest is not present anymore (`building levels` with lower priority is being shown instead). I did not upload the quests in the meantime (I did solve a few more quests though, which might have involved automatic or manual area downloads). Also, the SC might have been cleared from recent apps in the meantime.
![buildingtype_gone](https://user-images.githubusercontent.com/156656/133612325-f62c38cb-009b-4a04-bb52-2c934a01b932.jpg)


Also, trying to solve that quest on another phone (Android 6, SC 34.1) when I solve that particular building, the quest disappears normally.

**Comment by mnalis** [comment]
Anyway, unless there is reason to refrain from doing so (eg. it interferes with finding a root cause), I plan to upload those quests tomorrow.

**Comment by westnordost** [comment]
Uploading wll be fine. If there are duplicate changes, the duplicate ones will simply be dropped.

**Comment by westnordost** [comment]
The database is locked when some other process is updating it (such as during download), so at some point during download, solving quests will feel unresponsive/laggy for a short moment. But this is not the case here, as you can solve other quests.

So, right now I have no idea why this is happening. As this seems to be reproducible at least in one session and it is a debuggable version, I think it would be best that if that bug appears next time, get the device to a computer and set a breakpoint, see what's happening.

**Comment by Helium314** [comment]
Yesterday encountered the same bug twice, in both cases it was when opening hours changed during resurvey. Same resurvey on other elements worked normally.
I didn't really think about this issue, or I would have done more testing (take logs, restart app,...).

Anyway, at least some information to add: hiding the quest was possible, and answering other quest on the same element worked normally.

**Comment by Cj-Malone** [comment]
I've just experienced it with the parking fee quest. I first answered the quest while searching for new quests, I can't remember if the quest disappeared while the download was on going, but it's constantly re appearing after. The original answer shows up in the undo history, but not subsequent answers.

Searching for new quests hid it. I have auto upload turned off, changes were not uploaded when the quest disappeared.

**Comment by westnordost** [comment]
Sounds like a race-condition between solving a quest and downloading new quests. Something that is currently not synchronized should be synchronized.

**Comment by timothywashere** [comment]
As a heavy user, I have experienced this on many different types of quests but mostly road quests
I usually just restart the app and it's ok.
I always have my changes saved until manual upload.
Sometimes mobile data is on, sometimes off.

**Comment by mnalis** [comment]
I'm had it again, and managed perhaps to capture some more data. This time on address quest, trying to answer there is no housenumber. It simply does not even create an answer (even if it does animation like it did create answer). Forcing manual area scan didn't help either:

https://user-images.githubusercontent.com/156656/146596650-ffb1566a-2e63-4198-9770-c86b69af7169.mp4

Leaving a note finally removed the quest (at https://www.openstreetmap.org/note/2974158)

I've tried to capture the logcat when I got home for that time range (with `adb logcat --pid 31428 -b all`), but it doesn't seem to tell much (at least to me) at 15:20 when it happened (see video for time, it happens mostly in 15:20, and little in 15:21) :

```
12-17 15:19:16.515 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:19:16.956 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:19:18.567 31428 31428 I am_on_top_resumed_lost_called: [0,de.westnordost.streetcomplete.MainActivity,topStateChangedWhenResumed]
12-17 15:19:18.604 31428 31428 I am_on_paused_called: [0,de.westnordost.streetcomplete.MainActivity,performPause]
12-17 15:19:19.055 31428 31428 I am_on_stop_called: [0,de.westnordost.streetcomplete.MainActivity,STOP_ACTIVITY_ITEM]
12-17 15:19:20.432 31428 31428 I am_on_restart_called: [0,de.westnordost.streetcomplete.MainActivity,performRestartActivity]
12-17 15:19:20.475 31428 31428 I am_on_start_called: [0,de.westnordost.streetcomplete.MainActivity,handleStartActivity]
12-17 15:19:20.476 31428 31428 I am_on_resume_called: [0,de.westnordost.streetcomplete.MainActivity,RESUME_ACTIVITY]
12-17 15:19:20.482 31428 31428 I am_on_top_resumed_gained_called: [0,de.westnordost.streetcomplete.MainActivity,topWhenResuming]
12-17 15:19:24.241 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:19:24.667 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:19:37.807 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:19:42.882 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:19:43.372 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:20:03.613 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:20:04.153 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:20:10.650 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:20:11.058 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:20:20.965 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:20:21.424 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:20:25.436 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:20:30.879 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:20:31.419 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:20:35.713 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:20:36.188 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:20:43.655 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:20:45.123 31428 31428 I ViewRootImpl: dispatchDetachedFromWindow in doDie
12-17 15:21:09.429 31428 31428 I am_on_top_resumed_lost_called: [0,de.westnordost.streetcomplete.MainActivity,topStateChangedWhenResumed]
12-17 15:21:09.451 31428 31428 I am_on_paused_called: [0,de.westnordost.streetcomplete.MainActivity,performPause]
12-17 15:21:09.950 31428 31428 I am_on_stop_called: [0,de.westnordost.streetcomplete.MainActivity,STOP_ACTIVITY_ITEM]
12-17 15:21:36.827 31428 31428 I am_on_restart_called: [0,de.westnordost.streetcomplete.MainActivity,performRestartActivity]
12-17 15:21:36.854 31428 31428 I am_on_start_called: [0,de.westnordost.streetcomplete.MainActivity,handleStartActivity]
12-17 15:21:36.855 31428 31428 I am_on_resume_called: [0,de.westnordost.streetcomplete.MainActivity,RESUME_ACTIVITY]
12-17 15:21:36.859 31428 31428 I am_on_top_resumed_gained_called: [0,de.westnordost.streetcomplete.MainActivity,topWhenResuming]
12-17 15:21:51.736 31428 31428 I am_on_top_resumed_lost_called: [0,de.westnordost.streetcomplete.MainActivity,topStateChangedWhenResumed]
12-17 15:21:51.784 31428 31428 I am_on_paused_called: [0,de.westnordost.streetcomplete.MainActivity,performPause]
12-17 15:21:52.204 31428 31428 I am_on_stop_called: [0,de.westnordost.streetcomplete.MainActivity,STOP_ACTIVITY_ITEM]
12-17 15:21:53.076 31428 31428 I am_on_restart_called: [0,de.westnordost.streetcomplete.MainActivity,performRestartActivity]
12-17 15:21:53.114 31428 31428 I am_on_start_called: [0,de.westnordost.streetcomplete.MainActivity,handleStartActivity]
12-17 15:21:53.115 31428 31428 I am_on_resume_called: [0,de.westnordost.streetcomplete.MainActivity,RESUME_ACTIVITY]
12-17 15:21:53.122 31428 31428 I am_on_top_resumed_gained_called: [0,de.westnordost.streetcomplete.MainActivity,topWhenResuming]
12-17 15:21:53.706 31428 31428 I am_on_top_resumed_lost_called: [0,de.westnordost.streetcomplete.MainActivity,topStateChangedWhenResumed]
12-17 15:21:53.759 31428 31428 I am_on_paused_called: [0,de.westnordost.streetcomplete.MainActivity,performPause]
12-17 15:21:54.189 31428 31428 I am_on_stop_called: [0,de.westnordost.streetcomplete.MainActivity,STOP_ACTIVITY_ITEM]
12-17 15:23:07.073 31428 31428 I am_on_restart_called: [0,de.westnordost.streetcomplete.MainActivity,performRestartActivity]

```

Does it help at all? Or is there a way to obtain more information?

**Comment by matkoniecz** [comment]
house from video is https://www.openstreetmap.org/way/650186433

I see nothing special about it :(

Though there is something weird happening here: when I scan there I do not get quest for this `building=house` building with nothing interesting in logs.

I am confused. Is it on edge of tiles or whatever edge condition mat happen in housenumber filtering?

**Comment by mnalis** [comment]
@matkoniecz as mentioned, I left a note https://www.openstreetmap.org/note/2974158 as an answer to the question, so I would expect housenumber quest is no longer asked on it (and the note I left does not end in `?` character, so unless you configured SC preference to show all notes, it probably won't show quest for `OSM Note`). 

I guess if you close the note it should show up as addressquest in SC again? 

But I doubt the issue is related to that specific OSM object id; in my previous encounters (noted at the beginning of this issue) it seemed more like some race condition in StreetComplete that triggers the bug (as trying to solve same object next day worked without the problem, and the object didn't change in the meantime).

**Comment by matkoniecz** [comment]
Argh, now I feel stupid again. I got confused by different map orientation, manage to find it using different method and not noticed note. Sorry for doing it again on your issue.

**Comment by westnordost** [comment]
I'll try to reproduce it with this house

**Comment by westnordost** [comment]
I can't reproduce it with this house.

**Comment by mnalis** [comment]
Would adding debug in the code help track it down? 

I can add it in my branch (if it would produce too much noise in production SC) and try to reproduce it again, if anyone has ideas what to log and where.

**Comment by westnordost** [comment]
Well I changed something for v38.2, so maybe wait until it is reproducable with that version

**Comment by mnalis** [comment]
It seems to not have helped...I've had similar issues (quest not being accepted) with (my fork of) `39.0-beta1`:
- when solving `house number` quest and answering `no housenumber` to convert it to `building=yes` - it refused to accept answer for about a minute (6-7 tries), and then when I tried to record video of it, is suddenly started working all by itself.
- it seems to happens with smoothness quest for others too (as it did for me in the past): https://github.com/streetcomplete/StreetComplete/issues/3636

**Comment by smichel17** [comment]
> when I tried to record video of it, is suddenly started working all by itself.

Do you remember if you closed SC when you were switching to the video recording app?

**Comment by timothywashere** [comment]
I've had the same issue for smoothness and surface type, including when changing surface type. If you keep trying after 6 or 7 answers it usually accepts it. I've tried recording it twice and it worked instantly, SC did not close and reopen when attempting to record using the quick settings Android 10 menu. 
I use manual upload of answers, sometimes with mobile data on sometimes off, sometimes with data saver on.

**Comment by westnordost** [comment]
No point in creating a video of it. I believe you. 

In offline mode (auto sync off), does the changes counter go up or does nothing happen? Does the log _"Solved a <quest type> quest: <changes>"_ appear in the log?

Alright, next step is to add some logs. @mnalis , you offered to add this in your custom version?

1. In  QuestController:L203, add `if (e == null) Log.d(TAG, "Element ${q.elementType.name}#${q.elementId} is null!") else Log.d(TAG, "Element ${e.type.name}#${e.id} v ${e.version} is not applicable to ${q.questTypeName}!")`
2. You could add a `Log.d("ElementEditsController", "Added edit for ${edit.questType.javaClass.simpleName}")` in ElementEditsController:L144

**Comment by timothywashere** [comment]
Well I managed to create a video anyway  
I had solved the quest once before starting this video
[https://youtu.be/dLYQTGexEs4](https://youtu.be/dLYQTGexEs4)

**Comment by mnalis** [comment]
@smichel17 I've tried few times to answer a quest, each time, there is animation like the quest is being solved, but no solved quest appear in undo buffer, and he quest icon remain the same asking for `house number`. I've used navigation-square double-tap to switch to last used app (OsmAnd) and back, and tried again, and it still didn't work. Then I pulled down the android drawer (if that is what it's called) from top of the screen, and selected android screen recording option - the SC remained in foreground that time (only obscured by drawer of course). When I tried to solve the `house number` quest then, it suddenly worked. But in previous cases starting screen recording the same way does not avoid the problem (See https://github.com/streetcomplete/StreetComplete/issues/3290#issuecomment-996983305 for example). I'm starting to have a deja-vu from https://github.com/streetcomplete/StreetComplete/issues/3312 that it might be something random android does depending on resource shortage or whatever.

@westnordost I was in "auto upload off" mode and counter didn't increase, nor did the quest appear in undo menu. As I did't manage to record the bug I didn't thought to check the logcat; will do it next time. I've now added that two debug lines in my fork https://github.com/mnalis/StreetComplete/commit/d21e1f1773bf9dcbd807bbb929be201463f921ec and will try to reproduce it and come back with logs when it happens next time.

**Comment by mnalis** [comment]
@timothywashere that's interesting - in your video the "solved quest count" is actually **_decreasing_** with each unsuccessfull attempt! I don't think I've noticed that when it happened to me (AFAIR it stayed at same number, but I might have not noticed it?), and it seems quite strange to me!

**Comment by westnordost** [comment]
@timothywashere  uhh,... the counter actually decreased... ?

**Comment by westnordost** [comment]
That's a good hint anyway... the only time the counter goes down is

1. an edit has been marked as synced. @timothywashere clearly has auto-sync off so we can exclude that possibility here. I checked the code paths, an edit is marked as synced only by the uploader.

2. an edit that hasn't been synced yet is deleted. This is the only other place where the counter goes down. This would explain why the quest doesn't go away. But what edit is deleted? Even if it is added and then for some reason deleted immediately, it shouldn't go down, it should go up, then down.

So, when is an edit deleted:
1. when it was synced but applying it failed (edit conflict). Doesn't apply here, nothing is synced.
2. when it was undone (synced or not synced) by user interaction. Doesn't apply here, nothing is synced.
3. when a new edit is added that is the exact reverse of a previously added edit that hasn't been synced yet. https://github.com/streetcomplete/StreetComplete/blob/9db2a2aa9ca03682e2af43081b8adb68fd0e6e5d/app/src/main/java/de/westnordost/streetcomplete/data/osm/edits/ElementEditsController.kt#L33-L47

This must be the crux. Only, I don't see anything immediately wrong with this code. Will need to investigate. Maybe that code could also be completely thrown away as it is an unneeded optimization.

**Comment by timothywashere** [comment]
It looks like I split the way before changing the surface type, maybe the new section has the reverse change from the other section?

![Screenshot_20220113-013207](https://user-images.githubusercontent.com/87654968/149249786-e91ed8cc-c199-49e3-9a13-a79046cf1761.png)

**Comment by mnalis** [comment]
I've managed to reproduce it. Before starting to video at 14:56, I've already tried to answer `Surface is different` to `smoothness` quest once, and it continued asking for smoothness, so I've started to record the video. It auto-recovers on third attempt in this example, but sometimes it takes more tries or even closing/reopening app.
After video recording was completed, I've manually started upload of quests.

https://user-images.githubusercontent.com/156656/149361360-2bad1877-5fe0-4a5b-a435-e36fb5738557.mp4

I notice I've also got a decreasing counter -- but now looking at my previous video https://github.com/streetcomplete/StreetComplete/issues/3290#issuecomment-996983305, it seems to be happening there also, but I've never noticed the decrease before, just that it wasn't increasing :smile: 

(adb logcat coming in next comment)

**Comment by mnalis** [comment]
Here is logcat for `pid 14544` from `14:56` until the `Starting upload` after the bug happened. 
It is with those two debug lines added: https://github.com/mnalis/StreetComplete/commit/d21e1f1773bf9dcbd807bbb929be201463f921ec (hopefully at right places)


```
01-13 14:56:01.928 14544 14544 I Choreographer: Skipped 1222 frames!  The application may be doing too much work on its main thread.
01-13 14:56:01.943 14544 14544 D ActivityThread: Won't deliver top position change in state=5
01-13 14:56:01.981 14544 14566 D ZrHung.AppEyeUiProbe: restart watching
01-13 14:56:01.982 14544 14544 D SensorManager: 0x7e31d31fd0 addFd fd=123
01-13 14:56:01.985 14544 14544 I QuestAutoSyncer: Checking whether to automatically download new quests at 45.7967213,15.9315360
01-13 14:56:01.999 14544 14587 D mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
01-13 14:56:02.001 14544  8293 I QuestAutoSyncer: All downloaded in radius of 286 meters around user (6 tiles)
01-13 14:56:02.001 14544 29695 D mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
01-13 14:56:02.001 14544 29695 D Tangram : TANGRAM map.cpp:210: resize: 1080 x 2340
01-13 14:56:02.012 14544 29695 W libEGL  : EGLNativeWindowType 0x7e31ae9010 disconnect failed
01-13 14:56:02.013 14544 29695 D mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
01-13 14:56:02.013 14544 29695 D Tangram : TANGRAM map.cpp:210: resize: 1080 x 2340
01-13 14:56:02.019 14544 14544 I HwViewRootImpl: removeInvalidNode all the node in jank list is out of time
01-13 14:56:02.028 14544 14544 D DecorView: showOrHideHighlightView: hasFocus=true; winMode=1; isMrgNull=true
01-13 14:56:02.028 14544 14544 W InputMethodManager: startInputReason = 1
01-13 14:56:02.342 14544 14544 I QuestAutoSyncer: Checking whether to automatically download new quests at 45.7949909,15.9310834
01-13 14:56:02.358 14544  7396 I QuestAutoSyncer: All downloaded in radius of 286 meters around user (4 tiles)
01-13 14:56:05.157 14544 14544 I HwViewRootImpl: removeInvalidNode all the node in jank list is out of time
01-13 14:56:10.399 14544 14544 W HiTouch_PressGestureDetector: Touch pointer move a lot. The moving distance of X is:1.1986694, limit is:60The moving distance of Y is:86.57605, limit is:60
01-13 14:56:11.516 14544 14544 W HiTouch_PressGestureDetector: Touch pointer move a lot. The moving distance of X is:6.4746094, limit is:60The moving distance of Y is:103.813965, limit is:60
01-13 14:56:12.126 14544 14544 V AudioManager: querySoundEffectsEnabled...
01-13 14:56:12.144 14544 14544 D InputEventReceiver: dispatchInputInterval 1000000
01-13 14:56:12.160 14544 14587 D mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
01-13 14:56:12.179 14544 14544 I HwViewRootImpl: removeInvalidNode all the node in jank list is out of time
01-13 14:56:12.193 14544 14544 D DecorView: showOrHideHighlightView: hasFocus=false; winMode=1; isMrgNull=true
01-13 14:56:13.981 14544 14544 V AudioManager: querySoundEffectsEnabled...
01-13 14:56:13.999 14544 14544 I DecorView[]:  old windowMode:0 new windoMode:1
01-13 14:56:14.012 14544 14544 D InputEventReceiver: dispatchInputInterval 1000000
01-13 14:56:14.023 14544  9044 D HiTouch_PressGestureDetector: onAttached, package=de.westnordost.streetcomplete.mn.debug, windowType=2, mHiTouchRestricted=false
01-13 14:56:14.034 14544 14587 D mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
01-13 14:56:14.340 14544 14544 I ViewRootImpl: dispatchDetachedFromWindow in doDie
01-13 14:56:14.342 14544 14587 W libEGL  : EGLNativeWindowType 0x7e31c32950 disconnect failed
01-13 14:56:14.342 14544 14587 D OpenGLRenderer: endAllActiveAnimators on 0x7e0d1fd300 (MenuPopupWindow$MenuDropDownListView) with handle 0x7e0f3bcd40
01-13 14:56:15.303 14544 14544 W Settings: Setting device_provisioned has moved from android.provider.Settings.Secure to android.provider.Settings.Global.
01-13 14:56:15.303 14544 14544 V HiTouch_HiTouchSensor: User setup is finished.
01-13 14:56:15.409 14544 14544 V AudioManager: querySoundEffectsEnabled...
01-13 14:56:15.410 14544 14544 I ViewRootImpl: dispatchDetachedFromWindow in doDie
01-13 14:56:15.416 14544 14587 W libEGL  : EGLNativeWindowType 0x7e31a3f490 disconnect failed
01-13 14:56:15.416 14544 14587 D OpenGLRenderer: endAllActiveAnimators on 0x7e0cae1d00 (RippleDrawable) with handle 0x7e0f3bcd40
01-13 14:56:15.428 14544 14544 D DecorView: showOrHideHighlightView: hasFocus=true; winMode=1; isMrgNull=true
01-13 14:56:15.431 14544  7396 D QuestController: Solved a AddPathSmoothness quest: DELETE "surface"="asphalt"
01-13 14:56:15.738 14544  7396 D OsmQuestController: Created AddPathSurface for WAY#145431811
01-13 14:56:15.740 14544  7396 I OsmQuestController: Created 1 quests for 1 updated elements in 0.0s
01-13 14:56:15.742 14544  7396 I OsmQuestController: Persisted 1 new and removed 0 already resolved quests in 0.0s
01-13 14:56:17.790 14544 14544 I HwViewRootImpl: removeInvalidNode all the node in jank list is out of time
01-13 14:56:18.446 14544 14544 W HiTouch_PressGestureDetector: Touch pointer move a lot. The moving distance of X is:7.2678833, limit is:60The moving distance of Y is:81.4823, limit is:60
01-13 14:56:20.206 14544 14544 V AudioManager: querySoundEffectsEnabled...
01-13 14:56:21.381 14544 14544 V InputMethodManager: dispatchInputEvent
01-13 14:56:21.466 14544 14544 V InputMethodManager: dispatchInputEvent
01-13 14:56:21.472 14544 14544 I DecorView[]:  old windowMode:0 new windoMode:1
01-13 14:56:21.485 14544 14544 D InputEventReceiver: dispatchInputInterval 1000000
01-13 14:56:21.488 14544  9054 D HiTouch_PressGestureDetector: onAttached, package=de.westnordost.streetcomplete.mn.debug, windowType=2, mHiTouchRestricted=false
01-13 14:56:21.496 14544 14587 D mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
01-13 14:56:21.518 14544 14544 D DecorView: showOrHideHighlightView: hasFocus=false; winMode=1; isMrgNull=true
01-13 14:56:22.403 14544 14544 W Settings: Setting device_provisioned has moved from android.provider.Settings.Secure to android.provider.Settings.Global.
01-13 14:56:22.403 14544 14544 V HiTouch_HiTouchSensor: User setup is finished.
01-13 14:56:22.522 14544 14544 V AudioManager: querySoundEffectsEnabled...
01-13 14:56:22.527 14544 14544 I ViewRootImpl: dispatchDetachedFromWindow in doDie
01-13 14:56:22.533 14544 14587 W libEGL  : EGLNativeWindowType 0x7e31a3f490 disconnect failed
01-13 14:56:22.533 14544 14587 D OpenGLRenderer: endAllActiveAnimators on 0x7df679fa00 (RippleDrawable) with handle 0x7e0bff5840
01-13 14:56:22.546 14544 14544 D DecorView: showOrHideHighlightView: hasFocus=true; winMode=1; isMrgNull=true
01-13 14:56:23.660 14544 14544 D DecorView: showOrHideHighlightView: hasFocus=false; winMode=1; isMrgNull=true
01-13 14:56:23.670 14544 14544 D RtgSchedIpcFile: setCommandByIoctl failed ret:-1, cmdid:20, errno:101
01-13 14:56:23.673 14544 14544 D RtgSched: resetRtgSchedHandle failed enable:0
01-13 14:56:25.325 14544 14544 D DecorView: showOrHideHighlightView: hasFocus=true; winMode=1; isMrgNull=true
01-13 14:56:25.802 14544 14544 D DecorView: showOrHideHighlightView: hasFocus=false; winMode=1; isMrgNull=true
01-13 14:56:25.803 14544 14544 D RtgSched: resetRtgSchedHandle failed enable:0
01-13 14:56:27.321 14544 14544 D DecorView: showOrHideHighlightView: hasFocus=true; winMode=1; isMrgNull=true
01-13 14:56:27.904 14544 14544 I InputTransport: Fun_filterMotionEvent reportMoveEvent!
01-13 14:56:27.942 14544 14544 W HiTouch_PressGestureDetector: Touch pointer move a lot. The moving distance of X is:0.6793213, limit is:60The moving distance of Y is:72.774414, limit is:60
01-13 14:56:28.939 14544 14544 I HwViewRootImpl: removeInvalidNode all the node in jank list is out of time
01-13 14:56:30.160 14544 14544 W HiTouch_PressGestureDetector: Touch pointer move a lot. The moving distance of X is:26.082092, limit is:60The moving distance of Y is:88.75464, limit is:60
01-13 14:56:30.891 14544 14544 I InputTransport: Fun_filterMotionEvent reportMoveEvent!
01-13 14:56:30.916 14544 14544 W HiTouch_PressGestureDetector: Touch pointer move a lot. The moving distance of X is:12.741211, limit is:60The moving distance of Y is:116.95007, limit is:60
01-13 14:56:31.541 14544 14544 V AudioManager: querySoundEffectsEnabled...
01-13 14:56:31.558 14544 14544 D InputEventReceiver: dispatchInputInterval 1000000
01-13 14:56:31.575 14544 14587 D mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
01-13 14:56:31.609 14544 14544 D DecorView: showOrHideHighlightView: hasFocus=false; winMode=1; isMrgNull=true
01-13 14:56:32.583 14544 14544 V AudioManager: querySoundEffectsEnabled...
01-13 14:56:32.592 14544 14544 I DecorView[]:  old windowMode:0 new windoMode:1
01-13 14:56:32.607 14544 14544 D InputEventReceiver: dispatchInputInterval 1000000
01-13 14:56:32.619 14544  9090 D HiTouch_PressGestureDetector: onAttached, package=de.westnordost.streetcomplete.mn.debug, windowType=2, mHiTouchRestricted=false
01-13 14:56:32.630 14544 14587 D mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
01-13 14:56:32.937 14544 14544 I ViewRootImpl: dispatchDetachedFromWindow in doDie
01-13 14:56:32.939 14544 14587 W libEGL  : EGLNativeWindowType 0x7e877dc890 disconnect failed
01-13 14:56:32.939 14544 14587 D OpenGLRenderer: endAllActiveAnimators on 0x7df23e3b00 (MenuPopupWindow$MenuDropDownListView) with handle 0x7e0bf26e00
01-13 14:56:34.044 14544 14544 W Settings: Setting device_provisioned has moved from android.provider.Settings.Secure to android.provider.Settings.Global.
01-13 14:56:34.044 14544 14544 V HiTouch_HiTouchSensor: User setup is finished.
01-13 14:56:34.156 14544 14544 V AudioManager: querySoundEffectsEnabled...
01-13 14:56:34.158 14544 14544 I ViewRootImpl: dispatchDetachedFromWindow in doDie
01-13 14:56:34.164 14544 14587 W libEGL  : EGLNativeWindowType 0x7e31620610 disconnect failed
01-13 14:56:34.164 14544 14587 D OpenGLRenderer: endAllActiveAnimators on 0x7de9901a00 (RippleDrawable) with handle 0x7e0bf26e00
01-13 14:56:34.190 14544 14544 D DecorView: showOrHideHighlightView: hasFocus=true; winMode=1; isMrgNull=true
01-13 14:56:34.192 14544  8575 D QuestController: Solved a AddPathSmoothness quest: DELETE "surface"="asphalt"
01-13 14:56:34.530 14544  8292 D OsmQuestController: AddMaxHeight requires surrounding map data to determine applicability to WAY#141053208
01-13 14:56:34.575 14544  8292 I MapDataController: Fetched 774 elements and geometries in 45ms
01-13 14:56:34.579 14544  8575 D OsmQuestController: Created AddRoadSurface for WAY#141053208
01-13 14:56:34.579 14544  8575 D OsmQuestController: Created AddWayLit for WAY#141053208
01-13 14:56:34.582 14544  8575 I OsmQuestController: Created 2 quests for 1 updated elements in 0.1s
01-13 14:56:34.584 14544  8575 I OsmQuestController: Persisted 2 new and removed 0 already resolved quests in 0.0s
01-13 14:56:36.693 14544 14544 I HwViewRootImpl: removeInvalidNode all the node in jank list is out of time
01-13 14:56:37.535 14544 14544 W HiTouch_PressGestureDetector: Touch pointer move a lot. The moving distance of X is:11.855469, limit is:60The moving distance of Y is:91.663086, limit is:60
01-13 14:56:38.028 14544 14544 W HiTouch_PressGestureDetector: Touch pointer move a lot. The moving distance of X is:4.1223755, limit is:60The moving distance of Y is:92.71619, limit is:60
01-13 14:56:38.786 14544 14544 V AudioManager: querySoundEffectsEnabled...
01-13 14:56:38.801 14544 14544 D InputEventReceiver: dispatchInputInterval 1000000
01-13 14:56:38.813 14544 14587 D mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
01-13 14:56:38.833 14544 14544 D DecorView: showOrHideHighlightView: hasFocus=false; winMode=1; isMrgNull=true
01-13 14:56:40.198 14544 14544 V AudioManager: querySoundEffectsEnabled...
01-13 14:56:40.207 14544 14544 I DecorView[]:  old windowMode:0 new windoMode:1
01-13 14:56:40.221 14544 14544 D InputEventReceiver: dispatchInputInterval 1000000
01-13 14:56:40.233 14544  9099 D HiTouch_PressGestureDetector: onAttached, package=de.westnordost.streetcomplete.mn.debug, windowType=2, mHiTouchRestricted=false
01-13 14:56:40.241 14544 14587 D mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
01-13 14:56:40.369 14544 14557 I mplete.mn.debu: Background concurrent copying GC freed 234810(8289KB) AllocSpace objects, 0(0B) LOS objects, 38% free, 38MB/62MB, paused 41us total 124.387ms
01-13 14:56:40.557 14544 14544 I ViewRootImpl: dispatchDetachedFromWindow in doDie
01-13 14:56:40.560 14544 14587 W libEGL  : EGLNativeWindowType 0x7e31a3d7d0 disconnect failed
01-13 14:56:40.560 14544 14587 D OpenGLRenderer: endAllActiveAnimators on 0x7de97af300 (MenuPopupWindow$MenuDropDownListView) with handle 0x7e017723c0
01-13 14:56:41.832 14544 14544 W Settings: Setting device_provisioned has moved from android.provider.Settings.Secure to android.provider.Settings.Global.
01-13 14:56:41.832 14544 14544 V HiTouch_HiTouchSensor: User setup is finished.
01-13 14:56:41.953 14544 14544 V AudioManager: querySoundEffectsEnabled...
01-13 14:56:41.956 14544 14544 I ViewRootImpl: dispatchDetachedFromWindow in doDie
01-13 14:56:41.960 14544 14587 W libEGL  : EGLNativeWindowType 0x7e01965650 disconnect failed
01-13 14:56:41.960 14544 14587 D OpenGLRenderer: endAllActiveAnimators on 0x7de6046100 (RippleDrawable) with handle 0x7e8760dfc0
01-13 14:56:41.972 14544  8292 D QuestController: Solved a AddPathSmoothness quest: DELETE "surface"="asphalt"
01-13 14:56:41.975 14544 14544 D DecorView: showOrHideHighlightView: hasFocus=true; winMode=1; isMrgNull=true
01-13 14:56:42.020 14544  8292 D ElementEditsController: Added edit for AddPathSmoothness
01-13 14:56:42.024 14544  8292 D OsmQuestController: Created AddPathSurface for WAY#43924352
01-13 14:56:42.026 14544  8292 I OsmQuestController: Created 1 quests for 1 updated elements in 0.0s
01-13 14:56:42.029 14544  8292 I OsmQuestController: Persisted 1 new and removed 1 already resolved quests in 0.0s
01-13 14:56:44.505 14544 14544 I HwViewRootImpl: removeInvalidNode all the node in jank list is out of time
01-13 14:56:46.794 14544 14544 W HiTouch_PressGestureDetector: Touch pointer move a lot. The moving distance of X is:6.8723145, limit is:60The moving distance of Y is:88.33984, limit is:60
01-13 14:56:49.198 14544 14544 V AudioManager: querySoundEffectsEnabled...
01-13 14:56:50.189 14544 14544 V AudioManager: querySoundEffectsEnabled...
01-13 14:56:50.198 14544  8292 D QuestController: Solved a AddPathSurface quest: ADD "surface"="paving_stones"
01-13 14:56:50.252 14544  8292 D ElementEditsController: Added edit for AddPathSurface
01-13 14:56:50.255 14544  8292 D OsmQuestController: Created AddPathSmoothness for WAY#43924352
01-13 14:56:50.258 14544  8292 I OsmQuestController: Created 1 quests for 1 updated elements in 0.0s
01-13 14:56:50.262 14544  8292 I OsmQuestController: Persisted 1 new and removed 1 already resolved quests in 0.0s
01-13 14:56:54.521 14544 14544 I HwViewRootImpl: removeInvalidNode all the node in jank list is out of time
01-13 14:56:56.923 14544 14544 W HiTouch_PressGestureDetector: Touch pointer move a lot. The moving distance of X is:8.115662, limit is:60The moving distance of Y is:67.36731, limit is:60
01-13 14:56:59.437 14544 14544 V AudioManager: querySoundEffectsEnabled...
01-13 14:57:00.005 14544 14544 V AudioManager: querySoundEffectsEnabled...
01-13 14:57:00.015 14544  8292 D QuestController: Solved a AddPathSmoothness quest: ADD "smoothness"="good"
01-13 14:57:00.072 14544  8292 D ElementEditsController: Added edit for AddPathSmoothness
01-13 14:57:00.076 14544  8292 I OsmQuestController: Created 0 quests for 1 updated elements in 0.0s
01-13 14:57:00.077 14544  8292 I OsmQuestController: Persisted 0 new and removed 1 already resolved quests in 0.0s
01-13 14:57:06.589 14544 14544 V AudioManager: querySoundEffectsEnabled...
01-13 14:57:06.601 14544  8292 I Upload  : Starting upload
01-13 14:57:07.069 14544  8292 I ChangesetManager: Created changeset #116108386
01-13 14:57:07.170 14544  8292 D ElementEditsUploader: Uploaded a UpdateElementTagsAction
01-13 14:57:07.348 14544  8575 D NoteEditsUploader: Uploaded a CREATE to 3009091 at 45.79841313347967, 15.929601870873437
01-13 14:57:07.499 14544  8292 I OsmQuestController: Created 0 quests for 4 updated elements in 0.0s
01-13 14:57:07.499 14544  8292 I OsmQuestController: Persisted 0 new and removed 0 already resolved quests in 0.0s

```

**Comment by westnordost** [comment]
I compared the video with the log and compressed it to

<table>
<tr><th>time</th><th>log</th></tr>
<tr><th colspan="2">1. User answers that the surface is wrong but the quest doesn't go away:</th></tr>
<tr><td>-10.431</td><td>QuestController: Solved a AddPathSmoothness quest: DELETE "surface"="asphalt"</td></tr>
<tr><td>-10.738</td><td>OsmQuestController: Created AddPathSurface for WAY#145431811</td></tr>
<tr><td>-10.740</td><td>OsmQuestController: Created 1 quests for 1 updated elements in 0.0s</td></tr>
<tr><td>-10.742</td><td>OsmQuestController: Persisted 1 new and removed 0 already resolved quests in 0.0s</td></tr>
<tr><th colspan="2">2. User answers that the surface is wrong but the quest still doesn't go away. The changes counter goes down:</th></tr>
<tr><td>9.192</td><td>QuestController: Solved a AddPathSmoothness quest: DELETE "surface"="asphalt"</td></tr>
<tr><td>9.530</td><td>OsmQuestController: AddMaxHeight requires surrounding map data to determine applicability to WAY#141053208</td></tr>
<tr><td>9.575</td><td>MapDataController: Fetched 774 elements and geometries in 45ms</td></tr>
<tr><td>9.579</td><td>OsmQuestController: Created AddRoadSurface for WAY#141053208</td></tr>
<tr><td>9.579</td><td>OsmQuestController: Created AddWayLit for WAY#141053208</td></tr>
<tr><td>9.582</td><td>OsmQuestController: Created 2 quests for 1 updated elements in 0.1s</td></tr>
<tr><td>9.584</td><td>OsmQuestController: Persisted 2 new and removed 0 already resolved quests in 0.0s</td></tr>
<tr><th colspan="2">3. User answers that the surface is wrong a third time and now it goes away. The changes counter goes up:</th></tr>
<tr><td>17.369</td><td>Background concurrent copying GC freed 234810(8289KB) AllocSpace objects, 0(0B) LOS objects, 38% free, 38MB/62MB, paused 41us total 124.387ms</td></tr>
<tr><td>17.972</td><td>QuestController: Solved a AddPathSmoothness quest: DELETE "surface"="asphalt"</td></tr>
<tr><td>17.020</td><td>ElementEditsController: Added edit for AddPathSmoothness</td></tr>
<tr><td>17.024</td><td>OsmQuestController: Created AddPathSurface for WAY#43924352</td></tr>
<tr><td>17.026</td><td>OsmQuestController: Created 1 quests for 1 updated elements in 0.0s</td></tr>
<tr><td>17.029</td><td>OsmQuestController: Persisted 1 new and removed 1 already resolved quests in 0.0s</td></tr>
<tr><th colspan="2">4. User answers that the surface is paving stones. The changes counter goes up::</th></tr>
<tr><td>25.198</td><td>QuestController: Solved a AddPathSurface quest: ADD "surface"="paving_stones"</td></tr>
<tr><td>25.252</td><td>ElementEditsController: Added edit for AddPathSurface</td></tr>
<tr><td>25.255</td><td>OsmQuestController: Created AddPathSmoothness for WAY#43924352</td></tr>
<tr><td>25.258</td><td>OsmQuestController: Created 1 quests for 1 updated elements in 0.0s</td></tr>
<tr><td>25.262</td><td>OsmQuestController: Persisted 1 new and removed 1 already resolved quests in 0.0s</td></tr>
<tr><th colspan="2">5. User answers that the smoothness is good. The changes counter goes up::</th></tr>
<tr><td>35.015</td><td>QuestController: Solved a AddPathSmoothness quest: ADD "smoothness"="good"</td></tr>
<tr><td>35.072</td><td>ElementEditsController: Added edit for AddPathSmoothness</td></tr>
<tr><td>35.076</td><td>OsmQuestController: Created 0 quests for 1 updated elements in 0.0s</td></tr>
<tr><td>35.077</td><td>OsmQuestController: Persisted 0 new and removed 1 already resolved quests in 0.0s</td></tr>
</table>

**Comment by westnordost** [comment]
Observation: For 1. and 2. (where it doesn't work), both the counter goes down and no actual edit is created. So it seems, **some** edit is deleted instead.

..... aaand, incidentally I found the issue right now by scanning over the lines cited in https://github.com/streetcomplete/StreetComplete/issues/3290#issuecomment-1011579308 . But regardless, I will first write a few tests to ensure all parts of it work correctly plus also write a test that reveals the very bug we have here.

**Comment by westnordost** [comment]
So the issue was a missing `&&`. The result was that if the changes done for an edit on **any other element** was the reverse of this edit, that older edit was deleted instead. So, contributors with StreetComplete (with auto-sync off) have lost one solved quest every time they happened upon this bug.

I fixed this issue, but then decided to also completely remove this logic.

**Comment by westnordost** [comment]
Thank you @mnalis , I wouldn't have found it without your help. In particular the video+log was what made me find the issue. 👍

**Comment by cloudrac3r** [comment]
Thank you so so so much westnordost for inspecting this and solving it, and explaining what the issue was! You are awesome ❤ and I can't wait to try the new version :)

**Comment by smichel17** [comment]
> and explaining what the issue was!

+1, your in-depth explanations of how the app works (usually in the context why a bug happened or why a feature is too much work to implement) are very appreciated. They were a significant part of what got me contributing code to StreetComplete.

When I have some idea of what to expect before poking around in the code, I feel more confident that I will be able to find what I'm looking for in a reasonable amount of time. That makes me less hesitant to start. This is true of more than only StreetComplete; @mpcjanssen's explanations were a big reason I started contributing to SimpleTask.
